### DISCUSSION TITLE
Autocommit back behavior in implicit transactions

### DISCUSSION THREAD

**Discussion by igornero** [sqlalchemy/sqlalchemy#9933] [Ideas]
Hi!

I would suggest the following behavior:
Commands executed outside of an explicit transaction end up implicitly starting a new transaction, right? However, it remains pending commit or rollback.

How about when I run a command outside a transaction, that command already has its transaction started (which already happens today) and also finished (something like "autocommit=True").

The above behavior would be parameterized, that is, it would not be standard. Even so as not to impact those who already use the "Commit as you go" concept and to assist those who would like to execute a command outside of a transaction and have it simply reflected in the database.

I know this is an old discussion, but perhaps with a simple return parameter ("autocommit") it will be possible to maintain the current behavior and enable greater simplicity in codes where I would like to just execute a query or a DML without having to worry about a transaction is pending.

Thanks in advance for the excellent library.

**Comment by zzzeek** [comment]
we just removed this parameter in SQLAlchemy 2.0 after a period of deprecation and before that many years of discouraging its use

https://docs.sqlalchemy.org/en/20/changelog/migration_20.html#library-level-but-not-driver-level-autocommit-removed-from-both-core-and-orm
https://docs.sqlalchemy.org/en/20/changelog/migration_20.html#autocommit-mode-removed-from-session-autobegin-support-added

in other recent discussions (unfortunately I can't find it at the moment) we discussed what are the *actual* use cases people have that makes them want "autocommit" and we found it's really that they want a `Session` open for a long running period of time with objects that remain unexpired but also not flushed, without a transaction pointing to the database.  So in those discussion(s) I've proposed a new method that indicates the `Session` should close out existing connectivity for its ongoing transactions, while still keeping its notion of "transaction in progress" open.  on next use of database connectivity, the `SessionTransaction` would reopen its connections until the method is called again.

**Reply by zzzeek** [reply]
not to mention for this use case in 99% of the time what you really want is [driver level autocommit](https://docs.sqlalchemy.org/en/20/orm/session_transaction.html#setting-isolation-for-individual-transactions).   that will keep your long running session free of any transactional state yet the Session will still intuitively be tracking a checked out database connection.

**Reply by zzzeek** [reply]
basically my answer here: https://github.com/sqlalchemy/sqlalchemy/discussions/6921#discussioncomment-5761837

**Reply by igornero** [reply]
Okay, great.

In this case, I will apply the following configuration:
autobegin = False
expire_on_commit=False

That way, no transaction will be started implicitly (and an exception will also be thrown if one tries to start automatically).

And also for the functions of my Service classes that return Model objects, other queries will not be triggered when the consumer reads the return due to the expire_on_commit setting.

To facilitate the management of explicit transactions, I will use the following function with context manager:

```python
@contextmanager
def verify_transaction(db_session: Session) -> None:
     in_transaction = db_session.in_transaction()
     if not in_transaction:
         db_session.begin()

     try:
         yield

         if not in_transaction:
             db_session.commit()
     except:
         if not in_transaction:
             db_session.rollback()

         raise
```

The reason for checking in_transaction() is to prevent calls to begin() when a transaction is already started (which throws an exception). I could always start a transaction dynamically passing the nested = True or False property according to the return of the in_transaction() function, however, I prefer the above approach: if the caller started a transaction, OK, he is the owner of the transaction, otherwise I I will be the owner of the transaction.

That way, some boirlerplate code is prevented:

```python
with verify_transaction(self.db_session):
   result = self.db_session.get(Rj, rj_id)
   return result
```

Thank you so much for your answer.

**Reply by zzzeek** [reply]
prior versions of SQLAlchemy did this sort of thing automatically using this concept called "subtransactions".  people generally didnt understand it and I saw it misused all over the place.   so while it's a little more work on your end to do things explicitly, it's probably best because its workings are transparent.
